### ISSUE TITLE
on-demand cluster discovery is `not` working when a virtualhost uses multiple cluster(s) 

### ISSUE THREAD

**Issue by liwenwu-amazon** [envoyproxy/envoy#24726]
*Title*: *on-demand cluster discovery is not working when a virtualhost uses multiple cluster(s)*

*Description*:

We are using [Envoy v1.23.0](https://github.com/envoyproxy/envoy/tree/v1.23.0) and on-demand cluster discovery feature.  And for following route config

```
{
 "resources": [
  {
   "@type": "type.googleapis.com/envoy.config.route.v3.RouteConfiguration",
   "name": "liwen-on-demand-route",
   "virtualHosts": [
    {
     "name": "2596996162/svc-e29b3352-a656-4c7d-bc3e-5381136df17f",
     "domains": [
      "test1-second-0.liwen.com",
     ],
     "routes": [
      {
       "match": {
        "path": "/Ready0",
       },
       "route": {
        "weightedClusters": {
         "clusters": [
          {
           "name": "cluster-ready0-name",
          }
         ],
        }
       }
      },
      {
       "match": {
        "path": "/Ready1",
       },
       "route": {
        "weightedClusters": {
         "clusters": [
          {
           "name": "cluster-ready1-name",
          }
         ],
        }
       }
      },
    ],
     "virtualClusters": [
      {
       "headers": [
        {
         "name": ":authority"
        }
       ],
       "name": "account.12341234"
      }
     ],
     "typedPerFilterConfig": {
      "envoy.filters.http.on_demand": {
       "@type": "type.googleapis.com/envoy.extensions.filters.http.on_demand.v3.PerRouteConfig",
       "odcds": {
        "source": {
         "apiConfigSource": {
          "apiType": "DELTA_GRPC",
          "transportApiVersion": "V3",
          "grpcServices": [
           {
            "envoyGrpc": {
             "clusterName": "control-plane-xds-server"
            }
           }
          ],
          "setNodeOnFirstMessageOnly": true
         },
         "initialFetchTimeout": "0s",
         "resourceApiVersion": "V3"
        }
       }
      }
     },
     "retryPolicy": {
      "numRetries": 0
     }
    },

```

*Repro steps*:

when i  do curl for `.../Ready0`, I can see envoy does "on-demand" cluster discovery workflow
```
# 1st time for virtual host test1-second-0.liwen.com/Ready0
curl  `[envoy-server-ipaddr]:10000/Ready0` -H "Host:test1-second-0.liwen.com"
```

But when I do curl for `.../Ready1`, i do `not` see envoy doing any "on-demand" cluster discovery workflow

```
# 2nd time for virtual host test1-second-0.liwen.com/Ready1
curl  `[envoy-server-ipaddr]:10000/Ready1` -H "Host:test1-second-0.liwen.com"
```

* Potential Root Causes*

Looks like [OdCdsApiImpl::updateOnDemand()](https://github.com/envoyproxy/envoy/blob/release/v1.23/source/common/upstream/od_cds_api_impl.cc#L90) is done per `virtualhost` (e.g. test1-second-0.liwen.com) `not` per `cluster` .

In my use case,  

*  `curl ... /Ready0`.  which is 1st time on `virtualhost`(test1-second-0.liwen.com),  the `status_` is  `StartStatus::NotStarted` and then I can following trace line
`odcds: starting a subscription with cluster name cluster-ready0-name`

* then `curl ../Ready1`, which is 2nd time on `virtualhost`(test1-second-0.liwen.com), the `status_` is `StartStatus::InitialFetchDone` and I can see following trace line
 `odcds: requesting for cluster name cluster-ready1-name`

Since the work did `not` add subscription with cluster name `cluster-ready1-name`,  there was `no` callback when `cluster-ready1-name` returns.   And it finally `timesout`. and I can see following traceline

```
[2022-12-31 06:54:13.541][50538][debug] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1264] cm odcds: on-demand discovery for cluster cluster-ready1-name timed out
```

In Summary. from envoy trace I can see

```
odcds: starting a subscription with cluster name cluster-ready-0-name
```

I am `NOT` see above trace line for `cluster-ready-1-name`

**Comment by doujiang24** [comment]
Maybe you want PerRoute odcds:
https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/on_demand/v3/on_demand.proto#extensions-filters-http-on-demand-v3-ondemandcds

But I do agree there might be a bug, when VHDS & ODCDS is using at the same time.
It just using the filter level config here:
https://github.com/envoyproxy/envoy/blob/release/v1.23/source/extensions/filters/http/on_demand/on_demand_update.cc#L125-L127

**Comment by liwenwu-amazon** [comment]
@doujiang24 thanks,  do you have an working example config?  I am following [this test-odcds example](https://github.com/kinvolk/test-odcds) 

Here is snippets of  our config

```
"resources": [
  {
   "@type": "type.googleapis.com/envoy.config.listener.v3.Listener",
   "name": "default_listener",
   "address": {
    "socketAddress": {
     "address": "::",
     "portValue": 10000
    }
   },
   "filterChains": [
    {
     "filterChainMatch": {
      "destinationPort": 10000,
    ...
     "filters": [
      {
       "name": "envoy.filters.network.http_connection_manager",
       "typedConfig": {
        "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
        "rds": {
         "configSource": {
          "apiConfigSource": {
           "apiType": "DELTA_GRPC",
           "transportApiVersion": "V3",
           "grpcServices": [
            {
             "envoyGrpc": {
              "clusterName": "control-plane-xds-server"
             }
            }
           ],
           "setNodeOnFirstMessageOnly": true
          },
          "initialFetchTimeout": "0s",
          "resourceApiVersion": "V3"
         },
         "routeConfigName": "liwen-on-demand-route"
        },
        "httpFilters": [
         {
          "name": "envoy.filters.http.on_demand",
          "typedConfig": {
           "@type": "type.googleapis.com/envoy.extensions.filters.http.on_demand.v3.OnDemand"
          } 
         },
       ...
        {
          "name": "envoy.filters.http.router",
          "typedConfig": {
           "@type": "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router"
          }
         }
        ],
```

**Comment by doujiang24** [comment]
```
              routes:
              - name: t1
                match: { prefix: "/t1" }
                route:
                  cluster_header: "Cluster-name"
                typed_per_filter_config:
                  envoy.filters.http.on_demand:
                    "@type": type.googleapis.com/envoy.extensions.filters.http.on_demand.v3.PerRouteConfig
                    odcds:
                      source:
                        resource_api_version: V3
                        api_config_source:
                          api_type: DELTA_GRPC
                          transport_api_version: V3
                          grpc_services:
                            envoy_grpc:
                              cluster_name: xds_cluster

              - name: default
                match: { prefix: "/" }
                route:
                  cluster_header: "Cluster-Name"
                typed_per_filter_config:
                  envoy.filters.http.on_demand:
                    "@type": type.googleapis.com/envoy.extensions.filters.http.on_demand.v3.PerRouteConfig
                    odcds:
                      source:
                        resource_api_version: V3
                        api_config_source:
                          api_type: DELTA_GRPC
                          transport_api_version: V3
                          grpc_services:
                            envoy_grpc:
                              cluster_name: xds_cluster
```

@liwenwu-amazon I tried this, and it works. Also, I think it's just a workaround, which means per odcds per cluster.

After a deeper look, I think the root cause might be here, and elsewhere.
https://github.com/envoyproxy/envoy/blob/release/v1.23/source/common/upstream/od_cds_api_impl.cc#L86

it missing `updateResourceInterest` while watching a new cluster, so, only one cluster takes effect.
I'll try to file a PR to fix it.

**Comment by htuch** [comment]
@adisuissa
